### ISSUE TITLE
`allOf` schema validation error object do not compliance with the documentation.

### ISSUE THREAD

**Issue by nicktrandafil** [Tencent/rapidjson#1616]
Having a value:
```
rapidjosn::Value value(1);
```

and a schema:
```
{
    "allOf": [
        { "type": "string" },
        { "type": "string", "minLength": 2 },
        { "type": "string", "pattern": "^z" }
    ]
}
```

`rapidjson::SchemaValidator` gives an error object:
```
{
  "type": [
    {
      "expected": [
        "string"
      ],
      "actual": "integer",
      "instanceRef": "#",
      "schemaRef": "#/allOf/0"
    },
    {
      "expected": [
        "string"
      ],
      "actual": "integer",
      "instanceRef": "#",
      "schemaRef": "#/allOf/1"
    },
    {
      "expected": [
        "string"
      ],
      "actual": "integer",
      "instanceRef": "#",
      "schemaRef": "#/allOf/2"
    }
  ]
}
```

Which is not the one is expected reading the documentation for the structure of the error object.

I would rather expect the same error structure that is geven for `oneOf` and `anyOf`:
```
{
  "anyOf": {
    "errors": [
      {
        "type": {
          "expected": [
            "string"
          ],
          "actual": "integer",
          "instanceRef": "#",
          "schemaRef": "#/anyOf/0"
        }
      },
      {
        "type": {
          "expected": [
            "string"
          ],
          "actual": "integer",
          "instanceRef": "#",
          "schemaRef": "#/anyOf/1"
        }
      },
      {
        "type": {
          "expected": [
            "string"
          ],
          "actual": "integer",
          "instanceRef": "#",
          "schemaRef": "#/anyOf/2"
        }
      }
    ],
    "instanceRef": "#",
    "schemaRef": "#"
  }
}
```

So either the documentation or implementation should be fixed.
In case if the documentation incorrect and the implementation is correct,
I wonder why there exists such a thing as 'error merging' solely for `allOf`?

**Comment by nicktrandafil** [comment]
I refer to the section *Schema violation reporting* of the [documentation](https://rapidjson.org/md_doc_schema.html):

> allOf, anyOf, and oneOf
    errors: required array of at least one object. There will be as many items as there are subschemas in the allOf, anyOf or oneOf schema keyword, respectively. Each item will be the error value produced by validating the instance against the corresponding subschema...

It is not true for `allOf`.

**Comment by smhdfdl** [comment]
@nicktrandafil I have noticed this also.  I think it must have been a deliberate change. If you think about it, the semantics of `allOf` mean that **all** failures have to be fixed for the document to be valid. So there is no real need to have the failures separated into their own `errors` array, they can just be merged.

**Comment by nicktrandafil** [comment]
@smhdfdl this makes sense, thank you for explanation.

**Comment by smhdfdl** [comment]
@nicktrandafil I have submitted a PR that improves the way that errors are gathered and reported by the validator in `schema.h`. As part of the PR I have changed the `allOf` code so that it matches the documentation. While `allOf` errors can be merged as described above and as implemented in the current `master` branch, my team felt it would be clearer to our users if the error messages made it explicit that the errors were in the context of an `allOf`.  PR is #1837.
